test/e2e/storage/non_csi_volumes.go

```diff
@@ -0,0 +1,758 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+)
+
+type testDriver interface {
+	createDriver()
+	cleanupDriver()
+	getDriverInfo() driverInfo
+	getVolumeSource(readOnly bool) *v1.VolumeSource
+	getPersistentVolumeSource(readOnly bool) *v1.PersistentVolumeSource
+}
+
+type driverInfo struct {
+	name        string
+	fsGroup     *int64
+	podSec      v1.PodSecurityContext
+	maxFileSize int64
+
+	isBlockSupported bool
+
+	f      *framework.Framework
+	config framework.VolumeTestConfig
+	node   v1.Node
+}
+
+var testDrivers = map[string]func(f *framework.Framework, config framework.VolumeTestConfig, node v1.Node) testDriver{
+	"NFS":                        initNFSDriver,
+	"GlusterFS":                  initGlusterFSDriver,
+	"iSCSI [Feature:Volumes]":    initISCSIDriver,
+	"Ceph-RBD [Feature:Volumes]": initRbdDriver,
+	"CephFS [Feature:Volumes]":   initCephFSDriver,
+	"hostpath":                   initHostpathDriver,
+	"hostpathSymlink":            initHostpathSymlinkDriver,
+	"emptydir":                   initEmptydirDriver,
+}
+
+var _ = utils.SIGDescribe("Non-CSI", func() {
+	f := framework.NewDefaultFramework("volumes")
+
+	var (
+		cs     clientset.Interface
+		ns     *v1.Namespace
+		node   v1.Node
+		config framework.VolumeTestConfig
+	)
+
+	BeforeEach(func() {
+		cs = f.ClientSet
+		ns = f.Namespace
+		nodes := framework.GetReadySchedulableNodesOrDie(f.ClientSet)
+		node = nodes.Items[rand.Intn(len(nodes.Items))]
+		config = framework.VolumeTestConfig{
+			Namespace:      ns.Name,
+			Prefix:         "volume",
+			ClientNodeName: node.Name,
+			ServerNodeName: node.Name,
+		}
+	})
+
+	for driverName, initDriver := range testDrivers {
+		curDriverName := driverName
+		curInitDriver := initDriver
+
+		Context(fmt.Sprintf("using driver: %s", curDriverName), func() {
+			var (
+				driver testDriver
+			)
+
+			BeforeEach(func() {
+				driver = curInitDriver(f, config, node)
+				driver.createDriver()
+			})
+
+			AfterEach(func() {
+				driver.cleanupDriver()
+			})
+
+			It("should be mountable", func() {
```

Comment:
Can we hide the actual test cases in the suite-specific test file (like what is being done for subpath)?